primer.md

```diff
@@ -238,23 +238,54 @@ there are errors during the processing of a CloudEvent, the software
 encountering the error is encouraged to use the normal protocol-level error
 reporting to report them.
 
-## Versioning of Attributes
-
-For certain CloudEvents attributes, the entity or data model referenced by its
-value might change over time. For example, `dataschema` might reference one
-particular version of a schema document. Often these attribute values will then
-distinguish each variant by including some version-specific string as part of
-its value. For example, a version number (`v1`, `v2`), or a date (`2018-01-01`)
-might be used.
-
-The CloudEvents specification does not mandate any particular pattern to be
-used, or even the use of version strings at all. This decision is up to each
-event producer. However, when a version-specific string is included, care should
-be taken whenever its value changes as event consumers might be reliant on the
-existing value and thus a change could be interpreted as a "breaking change".
-Some form of communication between producers and consumers should be established
-to ensure the event consumers know what possible values might be used. In
-general, this is true for all CloudEvents attributes as well.
+## Versioning of CloudEvents
+
+For many CloudEvents, the schema of the data within the event might
+change over time. The CloudEvents specification does not mandate any
+particular pattern to be used, or even the use or consideration of
+versioning at all. This decision is up to each event producer.
+
+However, event producers are encouraged to consider how they might
+evolve their schema without breaking consumers. Two specific
+context attributes are particularly important in this respect, but
+differ in expected usage: `type` and `dataschema`.
+
+The `type` attribute is expected to be the primary means by which
+consumers identify the type of event that they receive. This could
+be achieved by subscribing to a specific CloudEvent type, or by
+filtering all received CloudEvents by type locally. But consumers
+who have identified a CloudEvent type will generally expect the data
+within that type to only change in backwardly-compatible ways,
+unless clearly indicated otherwise. The precise meaning of
+"backwardly-compatible" will vary by data content type.
+
+The `dataschema` attribute is expected to be informational, largely
+to be used during development and by tooling that is able to provide
+diagnostic information over arbitrary CloudEvents with a data
+content type understood by that tooling.
+
+When a CloudEvent's data changes in a backwardly-compatible way, the
+value of the `type` attribute should generally stay the same. The
+value of the `dataschema` attribute may change, or it may stay the
```

Comment:
I would suggest `s/may change, or/should change, though/`.  The implication otherwise is that assistive tooling (e.g. IDE plugins) or event consumers might have to poll for schema content changes rather than be able to depend on the unique resource identifier that a URI can provide.  I appreciated we are not forcing implementer's hands but the use of semver minor and patch version designators in URIs seems like a good best practice to encourage.  As stated, the mays seem equivalent and the implication I take away (when placing a probably undue level of attention here) is that we don't have an opinion.  At least **I** have a strong preference for the former, I can only speak for myself.

Reply:
I'm more on the fence - partly aware that I'm pretty sure our `dataschema` attributes *aren't* versioned that way.

I've made it "should" but acknowledged that the alternative is simpler... along with the reasons why it's more consumer-friendly to version.

I deliberately *haven't* called out semver minor changes here, partly as I can imagine one simple schema for `dataschema` attributes being a link to a file on GitHub, including the commit. That still has the "cache-busting" property, but doesn't require every schema change to be considered as "minor or patch" based on whether it's just a refactoring/comment change, and doesn't require tagging etc.

As ever, feel free to push back - and thanks for the input.

Reply:
Should feels like a win and as much as I can reasonably ask.  Thank you.  I think there's a lot of room for a "latest"/"evergreen" link that one would find and use through the discover service/portal but have thought that a strictly versioned dataschema would be best for each event, if one is even sending the attribute.  I have to admit not having thought of using commits in the URI but I now in hindsight it seems indefensible.  :D